### DISCUSSION TITLE
Why is runInSeparateProces removed in v4?

### DISCUSSION THREAD

**Discussion by dingo-d** [lucatume/wp-browser#662] [General]
I was running some tests where I want to set, and then unset the `define('WP_ADMIN', true);` constant (to simulate being on the admin side), and noticed my lines not being covered, and the last test where this should be `false` failing.

When debugging the tests, I noticed that the 

```php
/**
 * @runInSeparateProces
 * @preserveGlobalState disabled
 */
```

Annotations are no longer working and then noticed in the changelog that this is removed in V4.

This is a super useful feature for WordPress, seeing how it messes around with globals. Any idea when this feature will be re-added in v4?

**Comment by lucatume** [comment]
I will provide an immediate fix and an explanation.

Fix - use the [uopz extension to define and undefined constants](https://www.php.net/manual/en/function.uopz-redefine.php).

Explanation - the way the feature was implemented was brittle and never really satisfied me. Different versions of PHPUnit would handle things differently and it was a real pain to debug. My idea is to make an API around what is, currently, a [test trait](https://github.com/lucatume/wp-browser/blob/2ab6ffacabe06f1a287788905cdb224e45951de4/tests/_support/Traits/LoopIsolation.php#L23); that function is used [over and over in the project tests](https://github.com/lucatume/wp-browser/blob/e27c142f2a6efd66423708d34e8fdef78614cbe5/tests/unit/lucatume/WPBrowser/Module/WPLoaderTest.php#L1720-L1725) to achieve the same effect: the code executed in the Closure passed to the `LoopIsolation::assertInIsolation` method is executed in a separate process and the result is returned.

To avoid delaying the release of version 4 further, I had resolved to remove the feature to readd it only if really requested.

Could you provide an example of the kind of code you're currently running that leverages the `@runInSeparateProcess` annotation?

**Comment by dingo-d** [comment]
The `uopz` is an extension that I'd have to download separately and prepare the CI for. Will have to take a closer look at it.

I'll also try to see about that test trait.

The thing that is bothering me mostly is that I have a test where I need to check if the user is on the admin or not, so I set `define('WP_ADMIN', true);`, which will set `is_admin()` function to return true (and I can check for my branches).

For example, I have a code:

```php

public function myCallback()
{
	if(is_admin()) {
		return;
	}

	// Do some logic only on the front.
}
```

So to cover the actual business logic of the method I need to set the `WP_ADMIN` constant to `false`. And to cover the `if` branch it needs to be `true`.

So far I've 'circumvented' this by placing that test at the end of the class so that other tests before it will have the `WP_ADMIN` set to `false`, but the main problem is that the test classes after the one where I've put this constant to `true`, will now have that constant also set to `true`, thereby polluting the rest of the tests basically.

I've confirmed that by running a debug on the test class that gets executed after the test where the `WP_ADMIN` is set to `true`, and I can see the `is_admin()` returning `true` when I evaluate it.

**Comment by dingo-d** [comment]
The `LoopIsolation` trait is specific for the `wp-browser` test as far as I can see. Do you have a plan to maybe expose something like that as a module or something that can be used in userland testing?

**Reply by lucatume** [reply]
You can check the project own workflow files to see how to setup the extension in CI. To set it up locally, install the `uopz` extension as any other extension, but take care to set `uopz.exit=1` to let termination functions work correctly. 

Re `LoopIsolation`: yes, the idea would be to expose an API to make its use convenient in testing. Codeception 5 removed what little support there was to support the `@runInSeparateProcess` annotation ([link](https://github.com/Codeception/Codeception/issues/6113)), so I will explore what venues are there to support it in the project.

**Comment by lucatume** [comment]
I'm closing this discussion since [the feature has been reintroduced.](https://wpbrowser.wptestkit.dev/extensions/IsolationSupport/)
